llvm-hs-pure/src/LLVM/IRBuilder/Module.hs

```diff
@@ -167,6 +168,62 @@ extern nm argtys retty = do
   let funty = ptr $ FunctionType retty argtys False
   pure $ ConstantOperand $ C.GlobalReference funty nm
 
+-- | An external variadic argument function definition
+externVarArgs 
+  :: MonadModuleBuilder m
+  => Name   -- ^ Definition name
+  -> [Type] -- ^ Parameter types
+  -> Type   -- ^ Type
+  -> m Operand
+externVarArgs nm argtys retty = do
+  emitDefn $ GlobalDefinition functionDefaults
+    { name        = nm
+    , linkage     = External
+    , parameters  = ([Parameter ty (mkName "") [] | ty <- argtys], True)
+    , returnType  = retty
+    }
+  let funty = ptr $ FunctionType retty argtys True
+  pure $ ConstantOperand $ C.GlobalReference funty nm
+
+-- | A global variable definition
+global
+  :: MonadModuleBuilder m
+  => Name       -- ^ Variable name
+  -> Type       -- ^ Type
+  -> C.Constant -- ^ Initializer
+  -> m Operand
+global nm ty initVal = do
+  emitDefn $ GlobalDefinition globalVariableDefaults
+    { name                  = nm
+    , LLVM.AST.Global.type' = ty
+    , linkage               = Weak
+    , initializer           = Just initVal
+    }
+  pure $ ConstantOperand $ C.GlobalReference (ptr ty) nm
+
+-- | Creates a series of instructions to generate a pointer to a string
+-- constant. Useful for making format strings to pass to `printf`, for example
+globalStringPtr
+  :: MonadModuleBuilder m
+  => String -- ^ The string to generate
+  -> Name   -- ^ The name by which to refer to it
+  -> m Operand
+globalStringPtr str nm = do
+  let asciiVals = map (fromIntegral . ord) str
+      llvmVals  = map (C.Int 8) (asciiVals ++ [0]) -- append null terminator
+      char      = IntegerType 8
+      charStar  = ptr char
+      charArray = C.Array char llvmVals
+  emitDefn $ GlobalDefinition globalVariableDefaults
+    { name                  = nm
+    , LLVM.AST.Global.type' = LLVM.AST.Typed.typeOf charArray
+    , linkage               = Private
```

Comment:
Similar question to the one for weak linkage: What is the reasoning behind making this `Private`?

Reply:
The reason that I had weak originally was that in the Kaleidescope tutorial, Stephen Diehl wrote something about global variables having weak linkage in the mutable variables chapter. However, changing the linkage doesn't seem to affect whether things compile or not. Up to you if you'd like everything to have uniform External linkage or conform to what clang does.

Reply:
So for string literals I think our situation is a bit different: In C string literals are anonymous so it doesn’t make sense to give them non-private linkage. But the `globalStringPtr` API explicitely requires a name. I would suggest that the API should either accept a name and use external linkage or produce a name automatically and make that private. I’m open to both but if you have no preference, then I would suggest we go with the former.

As for global variables, the difference in linkage is only relevant if you link multiple modules together. If you have weak linkage and you have two declarations of the same name, they will be merged. So if this is a global variable, it would be shared across modules. I don’t have particularly strong feelings here so it would be nice if @sdiehl could weigh in but generally I find external linkage to be less surprising so I’m slightly leaning toward that. Either way, it would probably be good to document the choice in the haddocks.

Reply:
It would maybe be possible to compromise for string literals; something like:
```Haskell
-- | Creates a series of instructions to generate a pointer to a string
-- constant. Useful for making format strings to pass to @printf@, for example
globalStringPtr
  :: MonadModuleBuilder m
  => String       -- ^ The string to generate
  -> Maybe Name   -- ^ If a name is given, then the generated entity will be linked 
                  -- externally. If not, then a name will be chosen for it 
                  -- automatically and it will be linked privately.
  -> m Operand
globalStringPtr str nm = do
  let asciiVals = map (fromIntegral . ord) str
      llvmVals  = map (C.Int 8) (asciiVals ++ [0]) -- append null terminator
      char      = IntegerType 8
      charStar  = ptr char
      charArray = C.Array char llvmVals
  nm' <- case nm of Just given -> return given; Nothing -> fresh
  emitDefn $ GlobalDefinition globalVariableDefaults
    { name                  = nm'
    , LLVM.AST.Global.type' = LLVM.AST.Typed.typeOf charArray
    , linkage               = Private
    , isConstant            = True
    , initializer           = Just charArray
    , unnamedAddr           = Just GlobalAddr
    }
  pure $ ConstantOperand $ C.BitCast (C.GlobalReference charStar nm') charStar
```
could work, except that the `MonadModuleBuilder` class doesn't provide a convenient name generator. If you're OK with adding a `MonadIRBuilder` class constraint to the signature, then this would work, but otherwise, I think that going with external linkage is the easiest thing to do. 

For global variables I have no preference whatsoever.

Reply:
For the purpose of least surprises making External linkage the default seems sensible.

Reply:
Ok. Done in db3f39e.